lib/propolis/src/hw/qemu/fwcfg.rs

```diff
@@ -1044,3 +1045,98 @@ mod test {
         assert_eq!(0, pio_read::<u32>(&dev, FW_CFG_IOP_DMA_HI));
     }
 }
+
+pub mod formats {
+    use super::Entry;
+    use crate::hw::pci;
+
+    /// Collect one or more device elections for use in generating a boot order
+    /// `fw_cfg` entry, suitable for consumption by OVMF bootrom.
+    pub struct BootOrder(Vec<String>);
+    impl BootOrder {
+        pub fn new() -> Self {
+            Self(Vec::new())
+        }
+
+        /// Add a generic disk
+        pub fn add_disk(&mut self, loc: pci::BusLocation) {
+            // The OVMF logic is looking for "scsi"
+            let pci_path = Self::format_pci(loc, "scsi");
+            self.0.push(format!("{pci_path}/disk@0,0"));
+        }
+
+        /// Add generic PCI device
+        ///
+        /// For example, one might add an entry for an ethernet NIC as such:
+        /// ```
+        /// # use propolis::hw::qemu::fwcfg::formats::BootOrder;
+        /// # use propolis::hw::pci::BusLocation;
+        /// # let mut bootorder = BootOrder::new();
+        /// # let bus_loc = BusLocation::new(1, 0).unwrap();
+        /// bootorder.add_pci(bus_loc, "ethernet");
+        /// ```
+        pub fn add_pci(&mut self, loc: pci::BusLocation, kind: &str) {
+            self.0.push(Self::format_pci(loc, kind));
+        }
+
+        /// Add an NVMe disk.  This assumes namespace 1, as our NVMe emulation
+        /// does not currently support multiple namespaces in a device.
+        pub fn add_nvme(&mut self, loc: pci::BusLocation, eui: u64) {
+            // The decoding in OVMF demands that the bootorder entry identify
+            // nvme devices as vendor=0x8086 and device=5845.  While that
```

Comment:
well, that's cool :| 

(it _might_ be worth including a link to the relevant code in OVMF, but it's certainly not a big deal to me)